### PR TITLE
Implement the scenario options window

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Not much to see here yet, just checking in.

### CODE REVIEW

File: src/openloco/graphics/gfx.h [hunk 1]
```diff
@@ -145,19 +145,27 @@ namespace openloco::gfx
         uint8_t colour,
         string_id stringId,
         const void* args);
-    void draw_string_494B3F(
+    void draw_string_495224(
         drawpixelinfo_t& dpi,
         int16_t x,
         int16_t y,
+        int16_t width,
```
Comment (marijnvdwerf) [inline src/openloco/graphics/gfx.h]: Fix the original call; width is calculated at `0x00445EB1`.

File: src/openloco/localisation/string_ids.h [hunk 2]
```diff
@@ -532,6 +539,26 @@ namespace openloco::string_ids
...
 
+    constexpr string_id tooltip_scenario_options = 1674;
+    constexpr string_id tooltip_scenario_challenge = 1675;
+    constexpr string_id tooltip_company_options = 1676;
+    constexpr string_id tooltip_financial_options = 1677;
+    constexpr string_id title_scenario_options = 1678;
+    constexpr string_id title_scenario_challenge = 1679;
+    constexpr string_id title_company_options = 1680;
+    constexpr string_id title_financial_options = 1681;
+    constexpr string_id maximum_competing_companies = 1682;
```
Comment (marijnvdwerf) [inline src/openloco/localisation/string_ids.h]:
```suggestion
    constexpr string_id max_competing_companies = 1682;
```

File: src/openloco/graphics/gfx.h (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openloco/graphics/gfx.h]: See #341.

File: src/openloco/ui/dropdown.cpp [hunk 3]
```diff
@@ -78,6 +78,11 @@ namespace openloco::ui::dropdown
         add(static_cast<uint8_t>(index), title, { l });
     }
 
+    int16_t get_argument_at(size_t index, uint8_t offset)
+    {
+        return *(int16_t*)&_dropdownItemArgs[index][offset * 2];
```
Comment (AaronVanGeffen) [inline src/openloco/ui/dropdown.cpp]: I think this is an ugly hack. Perhaps it's better to keep a local copy for the cargo dropdown.

**Comment by AaronVanGeffen** [comment]
To do:

- [ ] Implement scenario options tab.
- [ ] Implement (common) update event(s).
- [ ] Remove need for dropdown::get_argument_at

File: src/openloco/localisation/string_ids.h [hunk 4]
```diff
@@ -612,7 +624,44 @@ namespace openloco::string_ids
...
+    constexpr string_id delay_before_competing_companies_start_months = 1685;
+    constexpr string_id selection_of_competing_companies = 1686;
+    constexpr string_id preferred_intelligence = 1687;
+    constexpr string_id preferred_aggressiveness = 1688;
+    constexpr string_id preferred_competitiveness = 1689;
+    constexpr string_id preference_any = 1690;
+    constexpr string_id preference_low = 1691;
+    constexpr string_id preference_medium = 1692;
+    constexpr string_id preference_high = 1693;
+
```
Comment (marijnvdwerf) [inline src/openloco/localisation/string_ids.h]: Why the blank line? The numbers are sequential

Comment (AaronVanGeffen) [reply src/openloco/localisation/string_ids.h]: They probably weren't when either of the strings were introduced. I'll amend.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 5]
```diff
@@ -0,0 +1,1194 @@
...
+        // 0x0043F67C
+        static void on_dropdown(window* self, widget_index widgetIndex, int16_t itemIndex)
+        {
+            if (itemIndex == -1)
+                return;
+
+            switch (widgetIndex)
+            {
+                case widx::preferred_intelligence_btn:
+                {
```
Comment (AaronVanGeffen) [inline src/openloco/windows/ScenarioOptions.cpp]: Unnecessary braces.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 6]
```diff
@@ -0,0 +1,1194 @@
...
+
+        // 0x0043FD14
+        static void on_mouse_down(window* self, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
+            {
+                case widx::objective_type_btn:
+                {
+                    widget_t& target = self->widgets[widx::objective_type];
+                    dropdown::show(self->x + target.left, self->y + target.top, target.width() - 3, target.height(), self->colours[1], std::size(objectiveTypeLabelIds), 0x80);
```
Comment (marijnvdwerf) [inline src/openloco/windows/ScenarioOptions.cpp]: These should be `width() - 4` instead, as `right - left` is already `width - 1`

File: src/openloco/windows/ScenarioOptions.cpp [hunk 7]
```diff
@@ -0,0 +1,1194 @@
...
+                            *objectivePerformanceIndex = std::max<uint8_t>(*objectivePerformanceIndex - 5, scenario::min_objective_performance_index);
+                            break;
+
+                        case scenario::objective_type::cargo_delivery:
+                        {
+                            uint32_t newDeliveredCargoAmount = *objectiveDeliveredCargoAmount;
+                            newDeliveredCargoAmount -= 100;
+                            uint16_t divisor = 100;
+
+                            if (*_clickRepeatTicks >= 100)
```
Comment (marijnvdwerf) [inline src/openloco/windows/ScenarioOptions.cpp]: Can we change this to something that doesn't allow different branches to be reached? e.g.

```c
if (_clickRepeatTicks < 100) {
    // - 100
else if(_clickRepeatTicks <200) {
    // -1000
} else {
    //  -10000
}
```

File: src/openloco/windows/ScenarioOptions.cpp [hunk 8]
```diff
@@ -0,0 +1,1194 @@
...
+
+                        case scenario::objective_type::performance_index:
+                            *objectivePerformanceIndex = std::max<uint8_t>(*objectivePerformanceIndex - 5, scenario::min_objective_performance_index);
+                            break;
+
+                        case scenario::objective_type::cargo_delivery:
+                        {
+                            uint32_t newDeliveredCargoAmount = *objectiveDeliveredCargoAmount;
+                            newDeliveredCargoAmount -= 100;
+                            uint16_t divisor = 100;
```
Comment (marijnvdwerf) [inline src/openloco/windows/ScenarioOptions.cpp]:
```suggestion
                            uint16_t stepSize = 100;
```

File: src/openloco/windows/ScenarioOptions.cpp [hunk 9]
```diff
@@ -0,0 +1,1194 @@
...
+                    dropdown::show(self->x + target.left, self->y + target.top, target.width() - 3, target.height(), self->colours[1], numCargoObjects, 0x80);
+
+                    uint16_t ddIdx = 0;
+                    for (uint16_t cargoIdx = 0; cargoIdx < maxCargoObjects; cargoIdx++)
+                    {
+                        auto cargoObject = objectmgr::get<cargo_object>(cargoIdx);
+                        if (cargoObject == nullptr)
+                            continue;
+
+                        dropdown::add(ddIdx, string_ids::dropdown_stringid, { cargoObject->name, cargoIdx });
```
Comment (marijnvdwerf) [inline src/openloco/windows/ScenarioOptions.cpp]:
```suggestion
                        dropdown::add(ddIdx, string_ids::dropdown_stringid, cargoObject->name);
```

Clean up when removing the dropdown argument accessor

File: src/openloco/windows/ScenarioOptions.cpp [hunk 10]
```diff
@@ -0,0 +1,1194 @@
...
+                    {
+                        auto cargoObject = objectmgr::get<cargo_object>(cargoIdx);
+                        if (cargoObject != nullptr)
+                            numCargoObjects++;
+                    }
+
+                    widget_t& target = self->widgets[widx::objective_cargo];
+                    dropdown::show(self->x + target.left, self->y + target.top, target.width() - 3, target.height(), self->colours[1], numCargoObjects, 0x80);
+
+                    uint16_t ddIdx = 0;
```
Comment (marijnvdwerf) [inline src/openloco/windows/ScenarioOptions.cpp]:
```suggestion
                    uint16_t dropdownIndex = 0;
```

File: src/openloco/windows/ScenarioOptions.cpp [hunk 11]
```diff
@@ -0,0 +1,1194 @@
...
+
+            self->widgets[common::widx::caption].right = self->width - 2;
+
+            self->widgets[common::widx::panel].right = self->width - 1;
+            self->widgets[common::widx::panel].bottom = self->height - 1;
+        }
+
+        static void switchTab(window* self, uint8_t targetTab)
+        {
+            if (input::is_tool_active(self->type, self->number))
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: I still don't like the shorthand. Don't want to change.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 12]
```diff
@@ -0,0 +1,1194 @@
...
+        }
+    }
+
+    // 0x0043EE58
+    window* open()
+    {
+        auto window = WindowManager::bringToFront(WindowType::scenarioOptions, 0);
+        if (window != nullptr)
+        {
+            if (input::is_tool_active(window->type, window->number))
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: I still don't like the shorthand. Don't want to change.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 13]
```diff
@@ -0,0 +1,1194 @@
...
+            self->widgets[common::widx::frame].right = self->width - 1;
+            self->widgets[common::widx::frame].bottom = self->height - 1;
+
+            self->widgets[common::widx::caption].right = self->width - 2;
+
+            self->widgets[common::widx::panel].right = self->width - 1;
+            self->widgets[common::widx::panel].bottom = self->height - 1;
+        }
+
+        static void switchTab(window* self, uint8_t targetTab)
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: Cheers. I thought I'd changed the signature, but it turns out those changes were stashed at some point.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 14]
```diff
@@ -0,0 +1,1194 @@
...
+            widgets[widx::objective_cargo].type = widget_type::none;
+            widgets[widx::objective_cargo_btn].type = widget_type::none;
+            widgets[widx::time_limit_value].type = widget_type::none;
+            widgets[widx::time_limit_value_down].type = widget_type::none;
+            widgets[widx::time_limit_value_up].type = widget_type::none;
+
+            switch (*objectiveType)
+            {
+                case scenario::objective_type::company_value:
+                    *(uint32_t*)&*commonFormatArgs = *objectiveCompanyValue;
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: There is no semantic difference, in this case, but I'll change it.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 15]
```diff
@@ -0,0 +1,1194 @@
...
+                    widgets[widx::objective_value].text = string_ids::challenge_monetary_value;
+                    break;
+
+                case scenario::objective_type::vehicle_profit:
+                    *(uint32_t*)&*commonFormatArgs = *objectiveMonthlyVehicleProfit;
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: There is no semantic difference, in this case, but I'll change it.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 16]
```diff
@@ -0,0 +1,1194 @@
...
+                    widgets[widx::objective_value].text = string_ids::challenge_monetary_value;
+                    break;
+
+                case scenario::objective_type::performance_index:
+                    *(uint16_t*)&*commonFormatArgs = *objectivePerformanceIndex;
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: There is no semantic difference, in this case, but I'll change it.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 17]
```diff
@@ -0,0 +1,1194 @@
...
+                    widgets[widx::objective_value].text = string_ids::challenge_performance_index;
+                    break;
+
+                case scenario::objective_type::cargo_delivery:
+                    *(uint32_t*)&*commonFormatArgs = *objectiveDeliveredCargoAmount;
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: There is no semantic difference, in this case, but I'll change it.

File: src/openloco/windows/ScenarioOptions.cpp [hunk 18]
```diff
@@ -0,0 +1,1194 @@
...
+            if ((*objectiveFlags & scenario::objective_flags::be_within_top_three_companies) != 0)
+                self->activated_widgets |= 1 << widx::check_be_within_top_three_companies;
+
+            if ((*objectiveFlags & scenario::objective_flags::within_time_limit) != 0)
+            {
+                self->activated_widgets |= 1 << widx::check_time_limit;
+                widgets[widx::time_limit_value].type = widget_type::wt_17;
+                widgets[widx::time_limit_value_down].type = widget_type::wt_11;
+                widgets[widx::time_limit_value_up].type = widget_type::wt_11;
+                commonFormatArgs[3] = *objectiveTimeLimitYears;
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/windows/ScenarioOptions.cpp]: `string_id` is defined as an `int16_t`. Explicit casting makes little sense.

File: src/openloco/objects/interface_skin_object.h [hunk 19]
```diff
@@ -78,10 +78,46 @@ namespace openloco
...
+        constexpr uint32_t tab_cup_frame14 = 143;
+        constexpr uint32_t tab_cup_frame15 = 144;
+
         constexpr uint32_t tab_cogs_frame0 = 168;
         constexpr uint32_t tab_cogs_frame1 = 169;
         constexpr uint32_t tab_cogs_frame2 = 170;
         constexpr uint32_t tab_cogs_frame3 = 171;
+        constexpr uint32_t tab_scenario_details = 172;
+        constexpr uint32_t tab_company = 173;
+        constexpr uint32_t tab_companies = 174;
```
Comment (marijnvdwerf) [inline src/openloco/objects/interface_skin_object.h]: whitespace between consecutive numbers

Comment (AaronVanGeffen) [reply src/openloco/objects/interface_skin_object.h]: Ha, you got me.

File: src/openloco/localisation/string_ids.h [hunk 20]
```diff
@@ -612,7 +624,43 @@ namespace openloco::string_ids
...
+    constexpr string_id preference_low = 1691;
+    constexpr string_id preference_medium = 1692;
+    constexpr string_id preference_high = 1693;
+    constexpr string_id starting_loan = 1694;
+    constexpr string_id starting_loan_value = 1695;
+    constexpr string_id max_loan_size = 1696;
+    constexpr string_id max_loan_size_value = 1697;
+    constexpr string_id loan_interest_rate = 1698;
+    constexpr string_id loan_interest_rate_value = 1699;
+
```
Comment (marijnvdwerf) [inline src/openloco/localisation/string_ids.h]: empty line

File: src/openloco/scenario.h [hunk 21]
```diff
@@ -16,9 +16,56 @@ namespace openloco::scenario
         allow_industries_start_up = (1 << 1),
     };
 
+    enum objective_flags : uint8_t
+    {
+        flag_0 = (1 << 0),
+        be_top_company = (1 << 1),
```
Comment (duncanspumpkin) [inline src/openloco/scenario.h:22]: These are all off by one.

**Pull request merged**